hoomd/hpmc/ShapeMoves.h

```diff
@@ -0,0 +1,870 @@
+        if (typid >= this->m_ntypes)
+            {
+            throw std::runtime_error("Invalid particle type.");
+            }
+        return typid;
+        }
+
+    Scalar getMoveProbability()
+        {
+        return this->m_move_probability;
+        }
+
+    void setMoveProbability(Scalar move_probability)
+        {
+        this->m_move_probability = fmin(move_probability, 1.0);
+        }
+
+    virtual Scalar operator()(uint64_t timestep,
+                              const unsigned int& N,
+                              const unsigned int type_id,
+                              const typename Shape::param_type& shape_new,
+                              const Scalar& inew,
+                              const typename Shape::param_type& shape_old,
+                              const Scalar& iold)
+        {
+        Scalar newdivold = inew / iold;
+        return (Scalar(N) / Scalar(2.0)) * log(newdivold);
+        }
+
+    virtual Scalar computeEnergy(uint64_t timestep,
+                                 const unsigned int& N,
+                                 const unsigned int type_id,
+                                 const typename Shape::param_type& shape,
+                                 const Scalar& inertia)
+        {
+        return 0.0;
+        }
+
+    protected:
+    std::shared_ptr<IntegratorHPMCMono<Shape>> m_mc;
+    Scalar m_det_inertia_tensor; // determinant of the moment of inertia tensor of the shape
+    std::shared_ptr<SystemDefinition> m_sysdef;
+    unsigned m_ntypes;
+    std::vector<Scalar> m_volume;
+    Scalar m_move_probability;
+    std::vector<vec3<Scalar>> m_centroids;
+
+    }; // end class ShapeMoveBase
+
+template<typename Shape> class PythonShapeMove : public ShapeMoveBase<Shape>
+    {
+    public:
+    PythonShapeMove(std::shared_ptr<SystemDefinition> sysdef,
+                    std::shared_ptr<IntegratorHPMCMono<Shape>> mc)
+        : ShapeMoveBase<Shape>(sysdef, mc), m_num_params(0)
+        {
+        m_params.resize(this->m_ntypes);
+        m_params_backup.resize(this->m_ntypes);
+        this->m_det_inertia_tensor = 1.0;
+        }
+
+    void prepare(uint64_t timestep)
+        {
+        m_params_backup = m_params;
+        }
+
+    void update_shape(uint64_t timestep,
+                      Scalar& stepsize,
+                      const unsigned int& type_id,
+                      typename Shape::param_type& shape,
+                      hoomd::RandomGenerator& rng)
+        {
+        for (unsigned int i = 0; i < m_params[type_id].size(); i++)
+            {
+            Scalar a = fmax(-stepsize, -(m_params[type_id][i]));
+            Scalar b = fmin(stepsize, (1.0 - m_params[type_id][i]));
+            hoomd::UniformDistribution<Scalar> uniform(a, b);
+            Scalar r = hoomd::detail::generate_canonical<double>(rng);
+            Scalar x = (r < this->m_move_probability) ? uniform(rng) : 0.0;
+            m_params[type_id][i] += x;
```

Comment:
What is `m_params` meant to represent? some sort of statistics on move numbers?

Reply:
see https://github.com/glotzerlab/hoomd-blue/pull/989#discussion_r846669732